src/index.js

```diff
@@ -1,212 +1,169 @@
-    visit(tree, 'code', remarkVisitor(blockCode));
-
-    function inlineCode(node, {highlighter, theme}) {
-      const meta = node.value.match(/{:([a-zA-Z.-]+)}$/)?.[1];
-
-      if (!meta) {
-        return;
+export function prettyCode(options = {}) {
+  const {
+    theme,
+    tokensMap = {},
+    onVisitLine = () => {},
+    onVisitHighlightedLine = () => {},
+    onVisitHighlightedWord = () => {},
+  } = options;
+
+  return async (tree) => {
+    if (
+      theme == null ||
+      typeof theme === 'string' ||
+      theme?.hasOwnProperty('tokenColors')
+    ) {
+      if (!highlighterCache.has('default')) {
+        highlighterCache.set('default', await shiki.getHighlighter({theme}));
       }
-
-      node.type = 'html';
-      // It's a token, not a lang
-      if (meta[0] === '.') {
-        if (typeof theme === 'string') {
-          throw new Error(
-            'MDX Pretty Code: Must be using a JSON theme object to use tokens.'
+    } else if (typeof theme === 'object') {
+      // color mode object
+      for (const [mode, value] of Object.entries(theme)) {
+        if (!highlighterCache.has(mode)) {
+          highlighterCache.set(
+            mode,
+            await shiki.getHighlighter({theme: value})
           );
         }
-
-        const color =
-          theme.tokenColors.find(({scope}) =>
-            scope?.includes(tokensMap[meta.slice(1)] ?? meta.slice(1))
-          )?.settings.foreground ?? 'inherit';
-
-        return sanitizeHtml(
-          `<span data-mdx-pretty-code data-color="${color}" data-theme="${
-            highlighter.themeKey
-          }"><span>${node.value.replace(/{:[a-zA-Z.-]+}/, '')}</span></span>`,
-          sanitizeOptions
-        );
       }
-
-      const highlighted = highlighter.codeToHtml(
-        node.value.replace(/{:\w+}/, ''),
-        meta
-      );
-
-      const dom = new JSDOM(highlighted);
-      const pre = dom.window.document.querySelector('pre');
-
-      return sanitizeHtml(
-        `<span data-mdx-pretty-code data-theme="${highlighter.themeKey}">${pre.innerHTML}</span>`,
-        sanitizeOptions
-      );
     }
 
-    function blockCode(node, {highlighter, loadedLanguages}) {
-      const lang =
-        ignoreUnknownLanguage && !loadedLanguages.includes(node.lang)
-          ? 'text'
-          : node.lang;
+    visit(tree, 'element', (node, index, parent) => {
+      // Inline code
+      if (
+        (node.tagName === 'code' && parent.tagName !== 'pre') ||
+        node.tagName === 'inlineCode'
+      ) {
+        const value = node.children[0].value;
 
-      node.type = 'html';
+        if (!value) {
+          return;
+        }
 
-      const highlighted = highlighter.codeToHtml(node.value, lang);
+        // TODO: allow escape characters to break out of highlighting
+        const stippedValue = value.replace(/{:[a-zA-Z.-]+}/, '');
+        const meta = value.match(/{:([a-zA-Z.-]+)}$/)?.[1];
+        const isLang = meta[0] !== '.';
```

Comment:
There's no way to break out of inline highlighting if you want to literally use `{:js}` (encountered this on the website I was making for this plugin); I broke out by adding a zero-width/whitespace character at the end but that's a hack.